### DISCUSSION TITLE
display-resample 触发大量掉帧/延迟帧的条件

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#385] [mpv 通用]
_这个帖子不是在讨论Gsync/freesync的问题，gsync和display-resample冲突也是已知问题，详情请自行看FAQ。_

https://github.com/mpv-player/mpv/issues/11122
这个问题比我想象的要普遍严重的多。
原本以为是因为240hz的刷新率太高，而且比较接近23.976fps的整十倍，视觉效果不算差，所以忍忍就过去了。现在测个4k95hz也掉帧实在忍不住了，不使用 `display-resample`抖动晃得瞎眼，我也想不通理论上只差1hz的偏离，实际体感会差这么多。

测试命令：
```ps
./mpv --config=no --idle=yes --force-window --video-sync=display-resample
```
播片的时候监控 stats 的 `dropped / mistimed / delayed frames`

从我自己的测试来看大概有几个明显的影响因素：
- 垂直同步倍率：60fps的视频不丢，24fps明显丢帧
- 源缩放率：720p的丢帧率比4k视频高的多（我显示器4k）
- 处理器性能：这是废话，我的配置基本算拉满级别的7950x了，处理器越弱自然更惨

官方短期没有修复计划，而现在2k240hz的显示器早都被小厂干到1.5k价位以下了。。。
没有太好的解决方案，我只能使用默认的 `--video-sync=audio` 然后让gsync去匹配mpv

**Comment by natural-harmonia-gropius** [comment]
我感觉补帧补到120效果很好，如果有传统插帧(--interplation)的vf滤镜可以试试。

**Reply by hooke007** [reply]
试了试简单的 `vf="fps=fps=3*${container-fps}"` 掉帧也炸裂了，我更看不明白了。
<img width=400 src="https://github.com/hooke007/MPV_lazy/assets/41094733/d8b5565e-1a57-49a2-b25a-0ac7acd652ad">
我印象中以前没这么严重，怀疑是不是这几年哪里负优化了(madvr那边也看到几个新的presentation glitches暴增的案例，合理怀疑n卡驱动也有一份责任)

update: mac也一样炸裂
<img width=400 src="https://github.com/hooke007/MPV_lazy/assets/41094733/e5a60144-e5ae-458c-a336-9e49aea74182">

**Comment by snylonue** [comment]
240 Hz 的屏幕
切换 osd 的时候掉帧比较厉害(最大观察到 20 Hz)，放一会就稳定了(<1 Hz)

**Reply by hooke007** [reply]
在我这里是osd会略微加剧，为了严谨性测了不同方式，比如播片玩暂停再开stats，仍然是很惨烈的数据

**Comment by signin404** [comment]
该片源用display-resample在我电脑上狂掉帧
同分辨率帧数的另两个片源就正常
https://pan.baidu.com/s/1XwK3t_4pHT0ISLIbPplaRA?pwd=z4dr

```
D:\MPV\App>mpv --config=no --idle=yes --force-window --video-sync=display-resample
[input] No key binding found for key 'BACK'.
 (+) Video --vid=1 (wmv3 1920x1080 30.000fps)
 (+) Audio --aid=1 --alang=jpn (wmalossless 2ch 48000Hz)
AO: [wasapi] 48000Hz stereo 2ch float
VO: [gpu] 1920x1080 yuv420p
AV: 00:00:02 / 00:01:26 (2%) A-V:  0.013 DS: 1.969697/1 Dropped: 1
Invalid audio PTS: 2.388667 -> 1.279000

Audio/Video desynchronisation detected! Possible reasons include too slow
hardware, temporary CPU spikes, broken drivers, and broken files. Audio
position will not match to the video (see A-V status field).
Consider trying `--profile=fast` and/or `--hwdec=auto-safe` as they may help.

AV: 00:01:24 / 00:01:26 (98%) A-V:  0.007 DS: 1.87/1 Dropped: 86
Invalid audio PTS: 84.607667 -> 84.949000
AV: 00:01:26 / 00:01:26 (100%) A-V:  0.000 DS: 1.7/1 Dropped: 96

 (+) Video --vid=1 (wmv3 1920x1080 30.000fps)
 (+) Audio --aid=1 --alang=jpn (wmav2 2ch 48000Hz)
AO: [wasapi] 48000Hz stereo 2ch float
VO: [gpu] 1920x1080 yuv420p
AV: 00:01:43 / 00:01:43 (100%) A-V:  0.000 DS: 2/1 Dropped: 1

 (+) Video --vid=1 (wmv3 1920x1080 30.000fps)
 (+) Audio --aid=1 --alang=jpn (wmav2 2ch 48000Hz)
AO: [wasapi] 48000Hz stereo 2ch float
VO: [gpu] 1920x1080 yuv420p
AV: 00:03:13 / 00:03:13 (100%) A-V:  0.000 DS: 2/1 Dropped: 1
```

**Reply by hooke007** [reply]
加个 --aid=0 跳过音轨的解码，目测音轨有问题

**Reply by signin404** [reply]
有点怪 加--aid=0
第一次播放不会掉
第二次播放就会掉

第二次播放不会应用--aid=0…

```
D:\MPV\App>mpv --config=no --idle=yes --force-window --video-sync=display-resample --aid=0
 (+) Video --vid=1 (wmv3 1920x1080 30.000fps)
     Audio --aid=1 --alang=jpn (wmalossless 2ch 48000Hz)
VO: [gpu] 1920x1080 yuv420p
V: 00:01:26 / 00:01:26 (100%) DS: 2/1

 (+) Video --vid=1 (wmv3 1920x1080 30.000fps)
 (+) Audio --aid=1 --alang=jpn (wmalossless 2ch 48000Hz)
AO: [wasapi] 48000Hz stereo 2ch float
VO: [gpu] 1920x1080 yuv420p
AV: 00:00:02 / 00:01:26 (2%) A-V:  0.013 DS: 1.969231/1 Dropped: 1
Invalid audio PTS: 2.388667 -> 1.279000

Audio/Video desynchronisation detected! Possible reasons include too slow
hardware, temporary CPU spikes, broken drivers, and broken files. Audio
position will not match to the video (see A-V status field).
Consider trying `--profile=fast` and/or `--hwdec=auto-safe` as they may help.

AV: 00:01:24 / 00:01:26 (98%) A-V:  0.004 DS: 1.87/2 Dropped: 85
Invalid audio PTS: 84.607667 -> 84.949000
AV: 00:01:26 / 00:01:26 (100%) A-V:  0.000 DS: 1.7/2 Dropped: 95
```

**Reply by hooke007** [reply]
> 第二次播放不会应用--aid=0…

如果你要重播要用 `--aid=no` ，否则切换文件会重置序号

> 是display-resample容错率太低么

比默认低那确实。除开视频、音频，应用在特效复杂的字幕时也会加剧掉帧。

**Reply by rufengsuixing** [reply]
特殊场景下太占性能了，如果放4k，开rife，本来显卡接近满载，再刷到120hz，电脑扛不住就掉帧了

**Reply by hooke007** [reply]
display-resample在4k120这个条件目前测下来不开任何东西都会掉帧，不需要上压力了

**Comment by GundamSaber77** [comment]
大概率应该是驱动的锅，我两台机器，一台12700K+4090，另一台5800X+1660S。两台显示器都是设置4K120。 4090那台机器播放视频在video-sync=display-resample模式下mistimed / delayed frames数值爆炸，一集动画播完最后数值好几千。。。就一开始半分钟还能维持住，大概生成帧率能到119.78左右徘徊，30秒真男人后就一直下跌，最后稳在105帧左右。令我非常懊恼，一直以为是性能不够，我MPV配置就是mpv_lazy标准版默认配置，什么都没开。
 直到前两天我在1660S那台机器上发现相同视频1660S竟然能维持住120左右的帧率！！一集动画下来mistimed / delayed frames基本也就刚过百，非常稳定！ 我就知道肯定不是性能问题了。各种控制变量排除法最后发现是驱动问题，我的1660S因为基本不打游戏，驱动比较老，而4090由于要打游戏，驱动基本都是最新版。我就一个版本一个版本的回退驱动测试，因为4090毕竟还要打游戏，肯定不能用太老的驱动，最后测试下来目前**_537.58_**这版驱动比较稳，一集24分钟1080P新番，mistimed / delayed frames在100左右，有些只有几十，非常稳！！！ 也在这里做下汇报，各位也可以试试

**Reply by hooke007** [reply]
正常应该是三0，偶尔出现个位数波动。(mistimed/delayed 此二项不应该出现持续增加的现象)

**Reply by GundamSaber77** [reply]
不过现在这样我已经能接受了，可以放心开video-sync=display-resample，之前大几千不敢开，不开了心里又非常不爽，就像有个疙瘩在一样。确实不应该持续增加，我个人感觉还是效率问题，显示器设置到60就基本就没有波动，120目前这波动对我来说也萌大奶了。

**Comment by wathper** [comment]
N卡控制面板中，开启 G-SYNC 并设置“仅全屏时使用”时：
令 MPV 窗口化播放`video-sync=display-resample`仅切换播放状态时会触发个位数的延迟帧、错时帧
令 mpv 全屏播放时`video-sync=display-resample`触发大量的延迟帧、错时帧。
将 G-SYNC 禁用后：
mpv 全屏、窗口化`video-sync=display-resample`均仅切换播放状态时会触发个位数的延迟帧、错时帧。

“G-SYNC”主要作用于游戏。或许与 mpv 的`video-sync=display-resample`存在兼容性问题。
是否关闭比较好？

**Reply by hooke007** [reply]
faq我就写了这两不兼容

**Reply by ttbxbm** [reply]
我前几天刚好开了SYNC，捣弄Cocker的时候，不晓得是不是这样原因，我关闭试试

**Comment by zch2136** [comment]
我对此也测试了一段时间，不是驱动问题，同样系统及驱动版本配置下，自23年11月5日后，mpv就始终存在这样的问题，近期有些版本有所缓解，但始终存在，23年12月31日前同比较好状态，但始终不如11月5日前的。

**Reply by snylonue** [reply]
https://github.com/mpv-player/mpv/commits/master/?since=2023-11-01&until=2023-11-06

简单看了一下，感觉 fe0d2b4 有些可疑

**Reply by snylonue** [reply]
https://github.com/shinchiro/mpv-winbuild-cmake/issues/555
其他平台有问题吗

**Reply by hooke007** [reply]
都是次要的，我前面就测过了不是近期的问题

https://github.com/hooke007/MPV_lazy/discussions/385#discussioncomment-7524875

**Comment by Es4c** [comment]
虽然不知道是不是同一个问题，但是全屏模式导致大量frame drop的问题我这边可以通过gpu-api=vulkan解决，该方案来源于5年前reddit的帖子：https://www.reddit.com/r/mpv/comments/fvkglz/massive_lags_in_fullscreen/
，不知道会不会有别的副作用？
同样试过帖子里提到的d3d11-flip=no，没什么用
如果这俩不是一个问题那就当我没说....

**Reply by hooke007** [reply]
这个问题faq里写了

**Comment by thcehocai** [comment]
对比了几个版本的mpvnet和mpvlazy，发现使用较新的libmpv-2.dll就会有持续增加的delayed，而用1月份以前的就很正常，确实是负优化了可能。

**Comment by XIONGPEILIN** [comment]
amd 的 free snyc 也是有这个问题，我一般用mpv就直接在显示器里面把 free snyc/Gsnyc关掉就不掉帧了

**Reply by hooke007** [reply]
https://github.com/mpv-player/mpv/issues/8537 gsync是mpv issue tracker里的已知问题，这里说的不是gsync的问题
